boa/src/syntax/parser/mod.rs

```diff
@@ -83,834 +159,1428 @@ impl Parser {
+    fn read_statement(&mut self) -> ParseResult {
+        let tok = self.next_skip_lineterminator()?;
+
+        let mut is_expression_statement = false;
+        let stmt = match tok.kind {
+            TokenKind::Keyword(Keyword::If) => self.read_if_statement(),
+            TokenKind::Keyword(Keyword::Var) => self.read_variable_statement(),
+            TokenKind::Keyword(Keyword::While) => self.read_while_statement(),
+            TokenKind::Keyword(Keyword::For) => self.read_for_statement(),
+            TokenKind::Keyword(Keyword::Return) => self.read_return_statement(),
+            TokenKind::Keyword(Keyword::Break) => self.read_break_statement(),
+            TokenKind::Keyword(Keyword::Continue) => self.read_continue_statement(),
+            TokenKind::Keyword(Keyword::Try) => self.read_try_statement(),
+            TokenKind::Keyword(Keyword::Throw) => self.read_throw_statement(),
+            TokenKind::Keyword(Keyword::Switch) => self.read_switch_statement(),
+            TokenKind::Punctuator(Punctuator::OpenBlock) => self.read_block_statement(),
+            // TODO: https://tc39.es/ecma262/#prod-LabelledStatement
+            // TokenKind::Punctuator(Punctuator::Semicolon) => {
+            //     return Ok(Node::new(NodeBase::Nope, tok.pos))
+            // }
+            _ => {
+                self.step_back();
+                is_expression_statement = true;
+                self.read_expression_statement()
+            }
+        };
+
+        match self.next_if_skip_lineterminator(TokenKind::Punctuator(Punctuator::Semicolon)) {
+            Ok(true) | Err(ParseError::NormalEOF) => {}
+            Ok(false) => {
+                if is_expression_statement {
+                    match self.peek(0)?.kind {
+                        TokenKind::LineTerminator
+                        | TokenKind::Punctuator(Punctuator::CloseBlock) => {}
+                        _ => {
+                            return Err(ParseError::Unexpected(self.get_current_token()?, None));
                         }
-                        _ => break,
                     }
                 }
+            }
+            Err(e) => return Err(e),
+        }
+
+        stmt
+    }
+
+    /// <https://tc39.es/ecma262/#prod-SwitchStatement>
+    fn read_switch_statement(&mut self) -> ParseResult {
+        // TODO: Reimplement the switch statement in the new parser.
+        unimplemented!("Switch statement parsing is not implemented");
+    }
+
+    /// <https://tc39.es/ecma262/#sec-expression-statement>
+    fn read_expression_statement(&mut self) -> ParseResult {
+        self.read_expression()
+    }
+
+    /// <https://tc39.es/ecma262/#sec-break-statement>
+    fn read_break_statement(&mut self) -> ParseResult {
+        let tok = self.get_next_token()?;
+        match tok.kind {
+            TokenKind::LineTerminator
+            | TokenKind::Punctuator(Punctuator::Semicolon)
+            | TokenKind::Punctuator(Punctuator::CloseParen) => {
```

Comment:
Shouldn't this be `Punctuator::CloseBlock`?

```suggestion
            | TokenKind::Punctuator(Punctuator::CloseBlock) => {
```

Reply:
I think so, too.

Reply:
The only context were `break` is allowed, is in `{}` braces.